gio/src/datagram_based.rs

```diff
@@ -0,0 +1,339 @@
+                condition,
+                cancellable.as_ref(),
+                None,
+                priority,
+                move |_, condition| {
+                    let _ = send.take().unwrap().send(condition);
+                    glib::Continue(false)
+                },
+            )
+        }))
+    }
+
+    fn create_source_stream<C: IsA<Cancellable>>(
+        &self,
+        condition: glib::IOCondition,
+        cancellable: Option<&C>,
+        priority: glib::Priority,
+    ) -> Pin<Box<dyn Stream<Item = glib::IOCondition> + 'static>> {
+        let cancellable: Option<Cancellable> = cancellable.map(|c| c.as_ref()).cloned();
+
+        let obj = self.clone();
+        Box::pin(glib::SourceStream::new(move |send| {
+            let send = Some(send);
+            obj.create_source(
+                condition,
+                cancellable.as_ref(),
+                None,
+                priority,
+                move |_, condition| {
+                    if send.as_ref().unwrap().unbounded_send(condition).is_err() {
+                        glib::Continue(false)
+                    } else {
+                        glib::Continue(true)
+                    }
+                },
+            )
+        }))
+    }
+
+    fn condition_wait(
+        &self,
+        condition: glib::IOCondition,
+        timeout: Option<Duration>,
+        cancellable: Option<&impl IsA<Cancellable>>,
+    ) -> Result<(), glib::Error> {
+        unsafe {
+            let mut error = ptr::null_mut();
+            let is_ok = ffi::g_datagram_based_condition_wait(
+                self.as_ref().to_glib_none().0,
+                condition.into_glib(),
+                timeout
+                    .map(|t| t.as_micros().try_into().unwrap())
+                    .unwrap_or(-1),
+                cancellable.map(|p| p.as_ref()).to_glib_none().0,
+                &mut error,
+            );
+            assert_eq!(is_ok == glib::ffi::GFALSE, !error.is_null());
+            if error.is_null() {
+                Ok(())
+            } else {
+                Err(from_glib_full(error))
+            }
+        }
+    }
+
+    fn receive_messages<'v, V: IntoIterator<Item = &'v mut [&'v mut [u8]]>, C: IsA<Cancellable>>(
+        &self,
+        vectors: V,
+        flags: i32,
+        timeout: Option<Duration>,
+        cancellable: Option<&C>,
+    ) -> Result<Vec<InputMessage>, glib::Error> {
+        let mut msgs = smallvec::SmallVec::<[_; 4]>::new();
+        let mut meta = vectors
+            .into_iter()
+            .map(|v| {
+                msgs.push(unsafe {
+                    std::mem::MaybeUninit::<ffi::GInputMessage>::zeroed().assume_init()
+                });
+                let vs = v
+                    .iter_mut()
+                    .map(|v| ffi::GInputVector {
+                        buffer: v.as_mut_ptr() as *mut _,
+                        size: v.len(),
+                    })
+                    .collect::<Vec<_>>();
```

Comment:
This seems suboptimal. Every receive being a new heap allocation even if there's nothing to actually receive currently...

Maybe this can make use of smallvec or the API can be defined in a way that the caller already provides something that is ABI-compatible with the C type?

Reply:
Ah this already uses smallvec. The question about having the caller provide something ABI-compatible stays though :)